### ISSUE TITLE
BUG: A connection object may become corrupt and re-referenced to another connection (with PATCH)

### ISSUE THREAD

**Issue by HenrikBengtsson** [r-devel/r-dev-day#27]
# TL;DR

A connection "handle" might end up pointing to another connection risking corrupting data, e.g.

```sh
# Create an empty file
con1 <- file("1.txt", open = "w")
close(con1)

# Create another empty file
con2 <- file("2.txt", open = "w")

# Write to the first connect
# Expected: an "invalid connection" error
# Actual: "hello" is written to file "2.txt"
cat("hello\n", file = con1)
and there's a risk 
close(con2)
```

After this, file `2.txt` will hold `hello\n`;

```r
readLines("2.txt")
#> [1] "hello"
```

# Troubleshooting

This is because connection handles are just integer indices that get reused, e.g.

```r
> con1 <- file("1.txt", open = "w")
> close(con1)
> con2 <- file("2.txt", open = "w")
``

At this point `con1` points the same connection as `con2`;

```r
> str(con1)
 'file' int 3
 - attr(*, "conn_id")=<externalptr>

> str(con2)
 'file' int 3
 - attr(*, "conn_id")=<externalptr> 

> con1
A connection with                    
description "2.txt" 
class       "file"  
mode        "w"     
text        "text"  
opened      "opened"
can read    "no"    
can write   "yes"

> con2
A connection with                    
description "2.txt" 
class       "file"  
mode        "w"     
text        "text"  
opened      "opened"
can read    "no"    
can write   "yes" 
```

# Suggestion

Internally, R connections are represented by an C-level [`Rconn` struct](https://github.com/wch/r-source/blob/tags/R-3-5-1/src/include/R_ext/Connections.h#L50-L85).  This structure has a field [`void *id`](https://github.com/wch/r-source/blob/tags/R-3-5-1/src/main/connections.c#L620-L624) that initated at zero and then incremented by one (= +4 bytes) for each new connection created (via `init_con()`).

Now, it turns out that the value of the unique `id` field is recorded in and exposed to us via (non-documented) attribute [`conn_id`](https://github.com/wch/r-source/blob/tags/R-3-5-1/src/main/connections.c#L5176-L5177) through the `void *ex_ptr` field (e.g. https://github.com/wch/r-source/blob/tags/R-3-5-1/src/main/connections.c#L5439-L5440 for file connections).  For example,

```r
> str(fh)
 'file' int 3
 - attr(*, "conn_id")=<externalptr>
> attr(fh, "conn_id")
<pointer: 0x373>  ## <== unique 'id' value

> str(fh2)
 'file' int 3
 - attr(*, "conn_id")=<externalptr>
> attr(fh2, "conn_id")
<pointer: 0x37c>  ## <== unique 'id' value
```

In other words, thanks to attribute `conn_id` a `connection` object does indeed carry all information needed to uniquely identify the underlying connection.  It is just that the current R implementation does not make use of it (other than for [garbage collection](https://github.com/wch/r-source/blob/tags/R-3-5-1/src/main/connections.c#L132-L136)/[finalizers](https://github.com/wch/r-source/blob/tags/R-3-5-1/src/main/connections.c#L5457)).

# Patch

I've got a 2018 prototype where R internally asserts that each `connection` used refers to the same connection as it was originally created for by validating against the internal `id` field.  See https://github.com/HenrikBengtsson/Wishlist-for-R/issues/81 for that patch.

# See also

https://github.com/HenrikBengtsson/Wishlist-for-R/issues/81

**Comment by shikokuchuo** [comment]
I just took a quick look at https://github.com/HenrikBengtsson/Wishlist-for-R/issues/81. If this is the latest version of the code then I think we need to re-visit this as curl uses the "conn_id" attribute, so there probably would need to be changes either here or at curl. I can spend some time at the next R dev day if there's desire to take this forward.

**Comment by mmaechler** [comment]
With soon another year passed,.., why has this still not made to R's Bugzilla ?

well, it has yesterday, in two comments by [me](https://bugs.r-project.org/show_bug.cgi?id=19172#c4) and [Ivan](https://bugs.r-project.org/show_bug.cgi?id=19172#c5), the latter notably mentioning this open dev-day issue.

**Comment by HenrikBengtsson** [comment]
> With soon another year passed,.., why has this still not made to R's Bugzilla ?

I think the answer to that is that we let things mature outside before moving to BugZilla. At least I do that to not flood BugZilla with issue reports, but also to make it easier for others to contribute, because BugZilla is not the place for that.

> well, it has yesterday, in two comments by [me](https://bugs.r-project.org/show_bug.cgi?id=19172#c4) and [Ivan](https://bugs.r-project.org/show_bug.cgi?id=19172#c5), the latter notably mentioning this open dev-day issue.

Thanks for pushing this forward - it's important.
